File: FullFlowExamples/go/PaymentForm_Deposit_FutureChargeDate.go
package depositFutureDate

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/splitit/splitit.sdks.go"
)

var apiClient = splitit.NewSandboxAPIClient(
	"apiKey",
	"username",
	"password",
)

// TODO: Use as a handler for Checkout request
func InitiateInstallmentPlan(transactionValue float64) (err error) {
	ctx := context.TODO()

	// TODO: Customize the data
	initPaymentResp, _, err := apiClient.InstallmentPlanApi.InstallmentPlanInitiate(
		ctx,
		splitit.InitiateInstallmentPlanRequest{
			PlanData: &splitit.PlanData{
				AutoCapture:          true,
				Amount:               &splitit.MoneyWithCurrencyCode{transactionValue, "USD"},
				NumberOfInstallments: 3,
				RefOrderNumber:       "abc123",
				// Request 50% of the transaction as a first installment
				FirstInstallmentAmount: &splitit.MoneyWithCurrencyCode{0.5 * transactionValue, "USD"},
				// Delay first charge for a week
				FirstChargeDate: splitit.NewSplititTime(time.Now().Add(7 * 24 * time.Hour)),
			},
			PaymentWizardData: &splitit.PaymentWizardData{
				RequestedNumberOfInstallments: "2,3,4,5,6",
			},
			// Optional data to pre-fill the form
			BillingAddress: &splitit.AddressData{
				AddressLine: "260 Madison Avenue.",
				City:        "New York",
				State:       "NY",
				Country:     "USA",
				Zip:         "10016",
			},
			// Optional data to pre-fill the form
			ConsumerData: &splitit.ConsumerData{
				FullName:    "John Smith",
				Email:       "JohnS@splitit.com",
				PhoneNumber: "1-415-775-4848",
				CultureName: "en-us",
			},
			// After user successfully interacts with splitit.com they would be
			// redirected to provided Succeeded URL with InstallmentPlanNumber as
			// a parameter in GET request. It is required to continue the flow.
			RedirectUrls: &splitit.RedirectUrls{
				Succeeded: "http://localhost/Succeeded",
				Canceled:  "http://localhost/Canceled",
				Failed:    "http://localhost/Failed",
			},
		},
	)
	if err != nil {
		return fmt.Errorf("InstallmentPlanInitiate call failed: %w", err)
	}

	// TODO: Redirect customer to initPaymentResp.CheckoutUrl
	fmt.Printf("Go to %s\n\n", initPaymentResp.CheckoutUrl)
	return nil
}

// TODO: Use as a handler for Succeeded URL callback
func VerifyPayment(installmentPlanNumber string, storedTransactionValue float64) (err error) {
	ctx := context.TODO()

	resp, _, err := apiClient.InstallmentPlanApi.InstallmentPlanVerifyPayment(
		ctx,
		splitit.VerifyPaymentRequest{
			InstallmentPlanNumber: installmentPlanNumber,
		},
	)
	if err != nil {
		return fmt.Errorf("InstallmentPlanVerifyPayment call failed: %w", err)
	}
	if !resp.IsPaid {
		return errors.New("transaction wasn't payed")
	}

	if resp.OriginalAmountPaid == storedTransactionValue {
		// Transaction was successfull
		return nil
	}

	// Transaction amount was tampered with. Try to refund.
	_, _, err = apiClient.InstallmentPlanApi.InstallmentPlanCancel(
		ctx,
		splitit.CancelInstallmentPlanRequest{
			InstallmentPlanNumber:  installmentPlanNumber,
			RefundUnderCancelation: splitit.REFUNDUNDERCANCELATION_ONLY_IF_A_FULL_REFUND_IS_POSSIBLE,
		},
	)
	if err != nil {
		return fmt.Errorf(
			"transaction value was tampered with, but the cancellation refund failed: %w",
			err,
		)
	}
	return errors.New("transaction value was tampered with, payment was refunded")
}

File: FullFlowExamples/java/PaymentFormDeposit_FutureChargeDate.java
import com.splitit.*;
import com.splitit.sdk.model.*;
import java.math.BigDecimal;
import java.math.RoundingMode;
import org.threeten.bp.OffsetDateTime;


public class PaymentFormDeposit_FutureChargeDate {
    private ApiClient apiClient;
    public PaymentFormDeposit_FutureChargeDate(String username, String password, String apiKey){
        this.apiClient = new ApiClient()
            .sandbox(true)
            .username(username)
            .password(password)
            .apiKey(apiKey);
    }

    // Use as a handler for Checkout request
    public void initiateInstallmentPlan(BigDecimal transactionValue) throws ApiException{

        // TODO: Customize the data
        var resp = apiClient.getInstallmentPlanApi().installmentPlanInitiate(
            new InitiateInstallmentPlanRequest()
            .planData(
                new PlanData()
                .autoCapture(true)
                .amount(
                    new MoneyWithCurrencyCode()
                    .value(transactionValue)
                    .currencyCode("USD")
                )
                .numberOfInstallments(3)
                .refOrderNumber("abc123")
                // Request 50% of the transaction as a first installment
                .firstInstallmentAmount(
                    new MoneyWithCurrencyCode()
                    .value(transactionValue.divide(new BigDecimal(2), 2, RoundingMode.CEILING))
                    .currencyCode("USD")
                )
                // Delay first charge for a week
                .firstChargeDate(
                    OffsetDateTime.now().plusWeeks(1)
                )
            )
            .paymentWizardData(
                new PaymentWizardData()
                .requestedNumberOfInstallments("2,3,4,5,6")
            )
            // Optional data to pre-fill the form
            .billingAddress(
                new AddressData()
                .addressLine("260 Madison Avenue.")
                .city("New York")
                .state("NY")
                .country("USA")
                .zip("10016")
            )
            .consumerData(
                new ConsumerData()
                .fullName("John Smith")
                .email("JohnS@splitit.com")
                .phoneNumber("1-415-775-4848")
                .cultureName("en-us")
            )
            // After user successfully interacts with splitit.com they would be
            // redirected to provided Succeeded URL with InstallmentPlanNumber as
            // a parameter in GET request. It is required to continue the flow.
            .redirectUrls(
                new RedirectUrls()
                .succeeded("http://localhost/Succeeded")
                .canceled("http://localhost/Canceled")
                .failed("http://localhost/Failed")
            )
        );
        // TODO: Redirect customer to initPaymentResp.CheckoutUrl
        System.out.println("Go to " + resp.getCheckoutUrl());
        //System.out.println(result);
    }

    // Use as a handler for Succeeded URL callback
    public void verifyPayment(String installmentPlanNumber, BigDecimal storedTransactionValue) throws ApiException{
        var resp = apiClient.getInstallmentPlanApi().installmentPlanVerifyPayment(
            new VerifyPaymentRequest()
            .installmentPlanNumber(installmentPlanNumber)
        );
        if (!resp.isIsPaid()){
            System.out.println("Transaction wasn't payed");
            return;
        }

        if (storedTransactionValue.setScale(3, RoundingMode.HALF_UP).compareTo(resp.getOriginalAmountPaid()) == 0){
            System.out.printf("Transaction for %.2f successfully verified\n", storedTransactionValue);
            return;
        }

        // the amounts didn't match
        apiClient.getInstallmentPlanApi().installmentPlanCancel(
            new CancelInstallmentPlanRequest()
            .installmentPlanNumber(installmentPlanNumber)
            .refundUnderCancelation(
                RefundUnderCancelation.ONLYIFAFULLREFUNDISPOSSIBLE
            )
        );
        System.out.println("Transaction value was tampered with, payment was refunded");
    }
}
